### DISCUSSION TITLE
/cmd_vel timeout?

### DISCUSSION THREAD

**Discussion by slowrunner** [iRobotEducation/create3_docs#516] [Questions] [answered]
### How are you connecting to your Create 3?

Wi-Fi (ROS 2)

### Computer(s) Model(s) and Operating System(s)

Ubuntu 22.04 in Docker over Raspberry Pi 5 PiOS Bookworm

### Which version of ROS 2 is installed on your computer?

Humble

### Which firmware version is installed on your robot?

H.2.4

### Which RMW is your robot running?

FastRTPS

### Is the robot connected to a network? If so, what is the network type?

Yes: 2.4GHz Home WiFi 

### Are there multiple Create 3 robots connected to your network?

No

### Is multicast enabled?

Yes

### What is the Adapter Board's USB/BLE Toggle currently switched to?

Bluetooth (default)

### Describe your question.

Is there more documentation on the Create3 implementation of the /cmd_vel topic?

All I found was the [listing of the topic with its type](https://iroboteducation.github.io/create3_docs/api/ros2/#:~:text=/cmd_vel%20%5Bgeometry_msgs/msg/Twist%5D).

I understood the affect on /cmd_vel of /safety_override, hazards, and max_acceleration and max_speed, **but** I did not understand there is a timeout.  **Did I miss the discussion of the timeout somewhere in the docs?**

My program spins with a 10Hz timer, and it seems that republishing the desired /cmd_vel at this rate is sufficient for smooth motion, but I actually had to undo a bunch of "if already published, don't publish every cycle" code from my original design to keep performing the motion I desired.

**Comment by adnan-saood** [comment]
the /cmd_vel topic is subscribed to by the diff-drive-controller I presume, nothing specific to the create3.

**Reply by slowrunner** [reply]
> the /cmd_vel topic is subscribed to by the diff-drive-controller I presume, nothing specific to the create3. 

I don’t use a diff drive controller node,  only the Create3 /motion_control node is listening to /cmd_vel topic in my bot.  I haven’t tried to figure out if the timeout is one of its gettable parameters.

**Reply by adnan-saood** [reply]
I couldn't infer the answer from the docs or from anywhere else. In my application I assumed that the `/cmd_vel` topic was subscribed to by a diff-drive-controller with an update rate of 20Hz.

**Reply by alsora** [reply]
This is not correct.
The Create 3 doesn't use the diff-drive-controller package

**Reply by adnan-saood** [reply]
Thanks for the clarification, i assumed it did since the name match the def input for the controller.

**Comment by alsora** [comment, marked as answer]
The Create 3 `/motion_control` node executes a 20Hz loop where it checks for the last received message on the `/cmd_vel` topic and forwards it to the wheels.
The frequency of this loop is not a parameter.

As you noticed, the node has indeed a timeout duration of 0.5 seconds.
If the last `/cmd_vel` message is older than the timeout, then the message will be considered stale and the robot will stop until a new message arrives.
This timeout is currently exposed as a parameter named `wheels_stop_threshold_sec`.
This parameter is "read-only", i.e. after the application has started you can't change it, however it can be modified via the ROS 2 parameter yaml exposed by the webserver https://iroboteducation.github.io/create3_docs/webserver/application/#application-ros-2-parameters-file (remember to restart the application after modifying the parameter).

I'll add this information to the docs.
